src/mlpack/methods/cf/cf.hpp

```diff
@@ -258,22 +258,24 @@ class CF
   //! Cleaned data matrix.
   arma::sp_mat cleanedData;
 
-  /**
-   * Helper function to insert a point into the recommendation matrices.
-   *
-   * @param queryIndex Index of point whose recommendations we are inserting
-   *     into.
-   * @param pos Position in list to insert into.
-   * @param neighbor Index of item being inserted as a recommendation.
-   * @param value Value of recommendation.
-   */
-  void InsertNeighbor(const size_t queryIndex,
-                      const size_t pos,
-                      const size_t neighbor,
-                      const double value,
-                      arma::Mat<size_t>& recommendations,
-                      arma::mat& values) const;
-
+  //! Candidate represents a possible recommendation.
+  struct Candidate
+  {
+    //! Value of this recommendation.
+    double value;
+    //! Item of this recommendation.
+    size_t item;
+    //! Trivial constructor.
+    Candidate(double value, size_t item) :
+        value(value),
+        item(item)
+    {};
+    //! Compare the value of two candidates.
+    friend bool operator>(const Candidate& l, const Candidate& r)
+    {
+      return l.value > r.value;
+    };
+  };
```

Comment:
@rcurtin  , yes I have considered std::pair<> but I thought it would be clearer if we have specific structure for this (maybe value/item is clearer than first/second...). Of course, I can use std::pair<> if you think it would be better :)
However, I can see a problem with the relational operators of std::pair.
They compare the first element and, if these elements are equal, they compare the second element. So, I think we should set "first" to represent the value, and "second" to represent the index.
When we decide if inserting a new element to the list of candidates, we make a comparison with the top element: "if (c > pqueue.top())"
If we use a pair, it will insert c in the case where:

c.first == pqueue.top().first && c.second > pqueue.top().second

This is an unnecesary overhead, because we don't care on the order of the indexes, so we are inserting a new element with the same value but a greater index...

We could easily fix this by explicitly comparing the first element: "if (c.first > pqueue.top().first)"

However, I can see another possible disadvantage. Inside the implementation of the priority queue, if we use the std::pair relational operator, elements will be not only sorted by value but also by index. This can cause a bit more overhead when shifting down/up through the heap.

Maybe I became a bit obsessed with efficiency and this doesn't make a real difference...